src/conda_package_handling/streaming.py

```diff
@@ -13,30 +16,54 @@
 from . import exceptions
 
 
-def _extract(filename: str, dest_dir: str, components: list[str]):
-    """
-    Extract .conda or .tar.bz2 package to dest_dir.
-
-    If it's a conda package, components may be ["pkg", "info"]
-
-    If it's a .tar.bz2 package, components must equal ["pkg"]
-
-    Internal. Skip directly to conda-package-streaming if you don't need
-    exception compatibility.
-    """
-
+def _stream_components(
+    filename: str,
+    components: list[str],
+    dest_dir: str = "",
+) -> Generator[tuple[TarFile, TarInfo], None, None]:
     if str(filename).endswith(".tar.bz2"):
         assert components == ["pkg"]
 
     try:
         with open(filename, "rb") as fileobj:
             for component in components:
                 # will parse zipfile twice
-                stream = package_streaming.stream_conda_component(
+                yield package_streaming.stream_conda_component(
                     filename, fileobj, component=component
                 )
-                extract_stream(stream, dest_dir)
     except cps_exceptions.CaseInsensitiveFileSystemError as e:
         raise exceptions.CaseInsensitiveFileSystemError(filename, dest_dir) from e
     except (OSError, TarError, BadZipFile) as e:
         raise exceptions.InvalidArchiveError(filename, f"failed with error: {str(e)}") from e
+
+
+def _extract(filename: str, dest_dir: str, components: list[str]):
+    """
+    Extract .conda or .tar.bz2 package to dest_dir.
+
+    If it's a conda package, components may be ["pkg", "info"]
+
+    If it's a .tar.bz2 package, components must equal ["pkg"]
+
+    Internal. Skip directly to conda-package-streaming if you don't need
+    exception compatibility.
+    """
+    for stream in _stream_components(filename, components, dest_dir=dest_dir):
+        try:
+            extract_stream(stream, dest_dir)
+        except cps_exceptions.CaseInsensitiveFileSystemError as e:
+            raise exceptions.CaseInsensitiveFileSystemError(filename, dest_dir) from e
+        except (OSError, TarError, BadZipFile) as e:
+            raise exceptions.InvalidArchiveError(filename, f"failed with error: {str(e)}") from e
+
+
+def _list(filename: str, components: list[str], verbose=True):
+    memfile = io.StringIO()
+    for component in _stream_components(filename, components):
+        for tar, _ in component:
+            with redirect_stdout(memfile):
+                tar.list(verbose=verbose)
+            component.close()
```

Comment:
huh interesting

Reply:
I'll add a comment here too for clarity.

Reply:
This part of the code is tricky and not entirely satisfying, it tries to do an explicit close at the end of a stream otherwise I noticed I was running out of network handles or file handles.